File: examples/sample-app/composeApp/src/commonMain/kotlin/dev/uatu/sample/Format.kt
package dev.uatu.sample

import kotlin.math.absoluteValue

private const val CURRENCY = "$"
private val AMOUNT_REGEX = Regex("""^\d+(\.\d{1,2})?$""")

fun formatCents(cents: Long, signed: Boolean = false): String {
    val abs = cents.absoluteValue
    val dollars = abs / 100
    val rem = (abs % 100).toInt()
    val dollarsStr = withThousandsSeparator(dollars)
    val remStr = rem.toString().padStart(2, '0')
    val body = "$CURRENCY$dollarsStr.$remStr"
    return when {
        cents < 0 -> "-$body"
        signed && cents > 0 -> "+$body"
        else -> body
    }
}

private fun withThousandsSeparator(value: Long): String {
    val s = value.toString()
    if (s.length <= 3) return s
    val out = StringBuilder()
    var count = 0
    for (i in s.length - 1 downTo 0) {
        out.append(s[i])
        count++
        if (count == 3 && i != 0) {
            out.append(',')
            count = 0
        }
    }
    return out.reverse().toString()
}

fun parseCents(input: String): Long? {
    val trimmed = input.trim().replace(",", "")
    if (trimmed.isEmpty()) return null
    if (!AMOUNT_REGEX.matches(trimmed)) return null
    val dot = trimmed.indexOf('.')
    val whole: String
    val frac: String
    if (dot < 0) {
        whole = trimmed
        frac = ""
    } else {
        whole = trimmed.substring(0, dot)
        frac = trimmed.substring(dot + 1)
    }
    val fracPadded = (frac + "00").substring(0, 2)
    val wholeLong = whole.toLongOrNull() ?: return null
    val fracLong = fracPadded.toLongOrNull() ?: return null
    val total = wholeLong * 100 + fracLong
    if (total < 0) return null
    return total
}

fun signedAmount(t: Transaction): Long = if (t.type == TxnType.credit) t.amount else -t.amount

fun balanceOf(txns: List<Transaction>): Long = txns.sumOf { signedAmount(it) }

fun initialsOf(name: String): String {
    val parts = name.trim().split(Regex("\\s+")).filter { it.isNotEmpty() }
    if (parts.isEmpty()) return "?"
    if (parts.size == 1) return parts[0].take(2).uppercase()
    return (parts.first().first().toString() + parts.last().first().toString()).uppercase()
}

expect fun formatDate(epochMillis: Long): String

expect fun formatClock(epochMillis: Long): String

File: examples/folio-web/src/format.ts
import type { Transaction } from './db'

const CURRENCY = '$'

export function formatCents(
  cents: number,
  opts: { signed?: boolean } = {},
): string {
  const abs = Math.abs(cents)
  const dollars = Math.floor(abs / 100)
  const rem = abs % 100
  const dollarsStr = dollars.toLocaleString('en-US')
  const formatted = `${CURRENCY}${dollarsStr}.${rem.toString().padStart(2, '0')}`
  if (cents < 0) return `-${formatted}`
  if (opts.signed && cents > 0) return `+${formatted}`
  return formatted
}

export function parseCents(input: string): number | null {
  const trimmed = input.trim().replace(/,/g, '')
  if (trimmed === '') return null
  if (!/^\d+(\.\d{1,2})?$/.test(trimmed)) return null
  const [whole, frac = ''] = trimmed.split('.')
  const fracPadded = (frac + '00').slice(0, 2)
  const total = Number(whole) * 100 + Number(fracPadded)
  if (!Number.isFinite(total) || !Number.isInteger(total)) return null
  if (total > Number.MAX_SAFE_INTEGER) return null
  return total
}

export function signedAmount(t: Transaction): number {
  return t.type === 'credit' ? t.amount : -t.amount
}

export function balanceOf(txns: Transaction[]): number {
  let total = 0
  for (const t of txns) total += signedAmount(t)
  return total
}

export function initialsOf(name: string): string {
  const parts = name.trim().split(/\s+/).filter(Boolean)
  if (parts.length === 0) return '?'
  if (parts.length === 1) return parts[0].slice(0, 2).toUpperCase()
  return (parts[0][0] + parts[parts.length - 1][0]).toUpperCase()
}

const DATE_FMT = new Intl.DateTimeFormat('en-US', {
  month: 'short',
  day: 'numeric',
  hour: 'numeric',
  minute: '2-digit',
})

export function formatDate(ts: number): string {
  return DATE_FMT.format(new Date(ts))
}

const CLOCK_FMT = new Intl.DateTimeFormat('en-US', {
  hour: 'numeric',
  minute: '2-digit',
  hour12: false,
})

export function formatClock(ts: number): string {
  return CLOCK_FMT.format(new Date(ts))
}
